### DISCUSSION TITLE
Updates of AeroDyn input file (preferably for 4.0)

### DISCUSSION THREAD

**Discussion by ebranlard** [OpenFAST/openfast#1895] [General]
I'm suggesting the following new inputs for the main AeroDyn file

**Wake_Mod**  (NOTE: new input has an underscore)
```Wake_Mod     - Type of wake/induction model (switch) {0=none, 1=BEMT, 2=VC/NW/OLAF-BEM, 3=OLAF} ```
 The previous input `WakeMod` is removed, `Wake_Mod=2` does not mean DBEMT anymore.  `WakeMod=2` is a placeholder for future induction calculation method. 

**~~AFAeroMod~~**
Removed (see UA_Mod) below

**~~FrozenWake~~**
Removed (see DBEMT_Mod) below

**BEM_Mod** 
`BEM_Mod =1` : the old aerodyn BEM implementation using the NoSweepPitchTwist coordinate system
`BEM_Mod =2` : the new aerodyn BEM implementation using the local staggered polar grid coordinate system.

NOTE: `BEM_Mod` currently governs the coordinate system used for ill-defined outputs such as the ones that ends with "x" and "y" (without specifying the coordinate system). Other ill-defined outputs are the typical BEM quantities such as "AxInd", "TnInd", "Phi", etc. These are defined in a different coordinate system depending on `BEM_Mod`. For consistency accross differents `Wake_Mod` (even when `Wake_Mod/=1`), we use `BEM_Mod` to determine the coordinate system of the ill-defined outputs. 

**DBEMT_Mod**
```DBEMT_Mod   - Type of dynamic BEMT (DBEMT) model {0=No Dynamic Wake, -1=Frozen Wake, 1=constant tau1, 2=time-dependent tau1, 3=constant tau1 with continuous formulation} (-) ```
 This falls under the other BEM options (i.e. only used when `WakeMod=1`). 
 The option `DBEMT_Mod=-1` has the same behavior as the `FrozenWake=True` and is meant for linearization. 
`DBEMT_Mod=0` has the same behavior as the previous `WakeMod=1` option, that is, no DBEMT. 
`DBEMT_Mod=J` (`J` in `1,2,3`)  , has the same behavior as the previous `WakeMod=2` & `DBEMT_Mod=J` 
(NOTE: not sure if we should have an underscore)

**Skew_Mod**
```Skew_Mod        - Select skew model {0: No skew model at all, -1:Throw away non-normal component, for linearization, 1: Active}```
Previous implementations always have the skew model on. This allows the user to turn it off completely. 
`Skew_Mod=-1` replaces the old `SkewMod=0` (an option that few users were using). This setting makes sure the wind speed is always normal to the rotor to limit period variation of the wind speed if the rotor is not perpendicular to the widn (e.g. tower top tilting or tilt). This is mostly needed for linearization. 
`Skew_Mod=1`, the Glauert model has two components: a momentum correction (not implemented before), and a velocity redistribution model (implemented before and referred to as the Pitt/Peters model). 

**SkewMomCorr**
```SkewMomCorr    - Turn the skew momentum correction on or off [used only when SkewMod=1]```
Previous behavior would be `SkewMomCorr=False`

**SkewRedistr_Mod**
```SkewRedistr_Mod Type of skewed-wake correction model (switch) {0: no redistribution, 1=Glauert/Pitt/Peters} [unsed only when Skew_Mod=1]```
Allows to turn on and off the redistribution model (couldn't be done previously), and give room for other redistributino model to be implemented (e.g. vortex cylinder model). 

**UA_Mod** (Now with an underscore)
```UA_Mod - Unsteady Aero Model Switch (switch) {0=Quasi-steady (no UA),  2=Gonzalez's variant (changes in Cn,Cc,Cm), etc...}```
Now introducing the `UA_Mod=0`, which turns off the unsteady aerodynamics (equivalent to the old `AFAeroMod=1`).
If `UA_Mod>0`, then this is equivalent to the old `AFAeroMod=2`. 

**AoA34** 
```AoA34  - Sample the AoA at the 3/4 chord or the aerodynamic center point {default=True} [always used]```
Determine whether the baseline angle of attack is sampled at the 3/4 chord or at the aerodynamic center point. 
Most UA_Mod will require `AoA34` to be set to true. But when using quasi steady aerodynamics, the user may want to set it to true or false. In the current AeroDyn, `AFAeroMod=1` implies `AoA34=False`. But to have a fair comparison with unsteady aerodynamics model, it would be best to set `AoA34=True` when when doing quasi-steady aero.

  | Old inputs               | Corresponding new inputs                                | 
  | ---                      | ---                                                     | 
  | `WakeMod=0`              | `Wake_Mod=0`                                            | 
  | `WakeMod=1` ("BEM")      | `Wake_Mod=1` and `DBEMT_Mod=0` and `BEM_Mod=1`          | 
  | `WakeMod=2` ("DBEMT")    | `Wake_Mod=1` and `DBEMT_Mod={1,2,3}`                    | 
  | `WakeMod=3` ("OLAF")     | `Wake_Mod=3`                                            | 
  | `AFAeroMod =1`           | `UA_Mod=0` and `AoA34=False`                            | 
  | `AFAeroMod =2`           | `UA_Mod>0` and `AoA34=False`  and `UA_Mod=UAMod`        | 
  | `FrozenWake=True`        | `DBEMT_Mod=-1`                                          | 
  | `FrozenWake=False`       | `DBEMT_Mod=0` (quasi-steady) or `DBEMT_Mod>0` (dynamic) | 
  | `SkewMod=2` (Glauert)    | `Skew_Mod=1` and `SkewRedistr_Mod=1`                    | 
  | `SkewMod=0` (Orthogonal) | `Skew_Mod=-1`                                           | 
  | `UAMod={2-7}`            | `UA_Mod={2-7}` and `AoA34=True`                         |

**Comment by ebranlard** [comment]
Here's how the input file would look like (I've commented the lines that are not changed):

```c
//------- AERODYN v15.03.* INPUT FILE ------------------------------------------------
//AeroDyn input file
//======  General Options  //============================================================================
//False                Echo        - Echo the input to "<rootname>.AD.ech"?  (flag)
//"default"            DTAero      - Time interval for aerodynamic calculations {or "default"} (s)
1                      Wake_Mod     - Type of wake/induction model (switch) {0=none, 1=BEMT, 3=OLAF} 
```
~~```0                       AFAeroMod- ```~~
```c
//0                    TwrPotent   - Type tower influence on wind based on potential flow around the tower (switch) {0=none, 1=baseline potential flow, 2=potential flow with Bak correction}
//0                    TwrShadow   - Calculate tower influence on wind based on downstream tower shadow (switch) {0=none, 1=Powles model, 2=Eames model}
//False                TwrAero     - Calculate tower aerodynamic loads? (flag)
```
~~```False                  FrozenWake  - ```~~
```c
//0                    TwrPotent   - Type tower influence on wind based on potential flow around the tower (switch) {0=none, 1=baseline potential flow, 2=potential flow with Bak correction}
//0                    TwrShadow   - Calculate tower influence on wind based on downstream tower shadow (switch) {0=none, 1=Powles model, 2=Eames model}
//False                TwrAero     - Calculate tower aerodynamic loads? (flag)
//False                CavitCheck  - Perform cavitation check? (flag) [AFAeroMod must be 1 when CavitCheck=true]
//False                Buoyancy    - Include buoyancy effects? (flag)
//False                CompAA      - Flag to compute AeroAcoustics calculation [only used when Wake_Mod=1]
//"unused"             AA_InputFile - AeroAcoustics input file [used only when CompAA=true]
//======  Environmental Conditions ===================================================================
//"default"            AirDens            - Air density (kg/m^3)
//"default"            KinVisc            - Kinematic viscosity of working fluid (m^2/s)
//"default"            SpdSound           - Speed of sound in working fluid (m/s)
//"default"            Patm               - Atmospheric pressure (Pa) [used only when CavitCheck=True]
//"default"            Pvap               - Vapour pressure of working fluid (Pa) [used only when CavitCheck=True]
//======  Blade-Element/Momentum Theory Options  ====================================================== [unused when Wake_Mod=0 or 3 except for BEM_Mod]
1                      BEM_Mod - Type of BEM model {1=legacy NoSweepPitchTwist, 2=polar} (switch) [used for all Wake_Mod to determine output coordinate system]
--- Skew
0                      Skew_Mod        - Select skew model {0: No skew model, -1:Remove non-normal component for linearization, 1: skew model active }
True                   SkewMomCorr    - Turn the skew momentum correction on or off [used only when Skew_Mod=1]
2                      SkewRedistr_Mod - Type of skewed-wake correction model (switch) {0=no redistribution, 1=Glauert/Pitt/Peters} [used only when Skew_Mod=1]
//"default"            SkewFactor     - Constant used in Pitt/Peters skewed wake model {or "default" is 15/32*pi} (-) [used only when SkewMod=2; unused when WakeMod=0 or 3]
--- BEM Algorithm [unused when Wake_Mod=0 or 3]
//True                 TipLoss     - Use the Prandtl tip-loss model? (flag)
//False                HubLoss     - Use the Prandtl hub-loss model? (flag)
//True                 TanInd      - Include tangential induction in BEMT calculations? (flag)
//True                 AIDrag      - Include the drag term in the axial-induction calculation? (flag)
//True                 TIDrag      - Include the drag term in the tangential-induction calculation? (flag) [unused when TanInd=FALSE]
//"default"            IndToler    - Convergence tolerance for BEMT nonlinear solve residual equation {or "default"} (-)
//100                  MaxIter        - Maximum number of iteration steps (-)
--- Shear [unused when Wake_Mod=0 or 3]
False                  SectAvg           - Use Sector Averaging (flag)
1                      SectAvgWeighting  - Weighting function for sector average  {1=Uniform}  within a 360/nB sector centered on the blade (switch) [used only when SectAvg=True] 
5                      SectAvgNPoints    - Number of points per sectors (-) [used only when SectAvg=True] 
--- Dynamic wake/inflow
0                      DBEMT_Mod    - Type of dynamic BEMT (DBEMT) model {0= No Dynamic Wake, -1= Frozen Wake for linearization, 1:constant tau1, 2=time-dependent tau1, 3=constant tau1 with continuous formulation} (-)
//29.0                 tau1_const  - Time constant for DBEMT (s) [used only when DBEMT_Mod=1]
//======  OLAF -- cOnvecting LAgrangian Filaments (Free Vortex Wake) Theory Options  ================== [used only when Wake_Mod=3]
//"unused"             OLAFInputFileName - Input file for OLAF [used only when Wake_Mod=3]
======  Unsteady Airfoil Aerodynamics Options  ====================================================
True                   AoA34       - Sample the angle of attack (AoA) at the 3/4 chord or the AC point {default=True} [always used]
0                      UA_Mod       - Unsteady Aero Model Switch (switch) {0=Quasi-steady (no UA),  2=Gonzalez's variant (changes in Cn,Cc,Cm), 3=Minnema/Pierce variant (changes in Cc and Cm)} 
True                   FLookup     - Flag to indicate whether a lookup for f' will be calculated (TRUE) or whether best-fit exponential equations will be used (FALSE); if FALSE S1-S4 must be provided in airfoil input files (flag) [used only when UA_Mod>0]

```

**Reply by jjonkman** [reply]
As we've discussed, I like it.  Thanks for sharing.

The only thing I would clarify is that `DBEMT_Mod` = -1 (which replaces `FrozenWake` = True) is meant for linearization, similar to `SkewMod` = -1.

**Reply by jjonkman** [reply]
Also, doesn't the "Shear" section also need an input associated with the number of sampled points within the sector?

**Reply by ebranlard** [reply]
It seems I had removed it, I've now put it back in

**Comment by mebrahiimi** [comment]
Hello,
I've been researching the cosine loss function, so I decided to simulate different yaw angle for IEA-15 MW bottom fixed with openFAST v4, the wind is steady, just the Generator DOF is activated, and I applied the yaw misalignment with changing in NacYaw. I activated the skewed correction in aerodyn, as here:
1             BEM_Mod            - BEM model {1=legacy NoSweepPitchTwist, 2=polar} (switch) [used for all Wake_Mod to determine output coordinate system]
--- Skew correction
1             Skew_Mod           - Skew model {0=No skew model, -1=Remove non-normal component for linearization, 1=skew model active}
True         SkewMomCorr        - Turn the skew momentum correction on or off [used only when Skew_Mod=1]
default       SkewRedistr_Mod    - Type of skewed-wake correction model  { 1=Glauert/Pitt/Peters, default=1} [used only when Skew_Mod=1]
default     SkewRedistrFactor  - Constant used in Pitt/Peters skewed wake model {or "default" is 15/32*pi} (-) [used only when Skew_Mod=1 and SkewRedistr_Mod=1]
As can be seen, I plotted the results for wind = 6 m/s, and I plotted with cosine power coefficient for 1.88, and 3 [P(yaw) = cos(yaw)^n, n: cosine power coefficient]. I believed if the skewed correction works, it should deviate from the coefficient of 3, but it's not. could you enlighten me with some idea.
Thank you
![wind6](https://github.com/user-attachments/assets/3189c6ca-40b5-4ce0-b0b4-8a84454e5b11)

**Reply by jjonkman** [reply]
Dear @majidpoli,

To use the new BEM method valid for high skew that is documented in our TORQUE 2024 paper (https://www.nrel.gov/docs/fy24osti/88162.pdf), you should use `SkewMomCorr` = TRUE (as you've done) along with `BEM_Mod` = 2 (which you've set to 1).  See the updated AeroDyn input file documentation for more information: https://openfast.readthedocs.io/en/main/source/user/aerodyn/input.html#blade-element-momentum-theory-options.

Best regards,

**Comment by mebrahiimi** [comment]
Thank you, I had some try with IEA-15MW to understand the cosine power and thrust coefficient, I flagged just CompElast:1, CompInflow:1, CompAero: 2 and servodyn = 1, total sim time = 300 - (50s).
* wind: 5,7,9,11 with shear = 0, type: steady
* 2             BEM_Mod            - BEM model   
  1             Skew_Mod           - Skew model 
True         SkewMomCorr        
default       SkewRedistr_Mod    
default     SkewRedistrFactor  
* I set shafttilt = 0, and all initial 
conditions to zero, but just rotorpeed = 5.7 rpm,
I can see some differences in the power coefficient in comparison with the paper, (for power: 1.75, for thrust: 0.64), 
I have Power: 1.15, thrust : 0.8.
am I missing something?
![fit_function](https://github.com/user-attachments/assets/37817747-d7a2-4162-97a6-1903b30a3bcb)
![fit_function_thrust](https://github.com/user-attachments/assets/97c7b37b-269d-4100-80f2-98c07ea5e8c8)
